[dist_optim] add distributed functional sgd optimizer

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50630 [dist_optim] add warning to distributed optimizer
* #50620 [dist_optim] add distributed functional AdamW optimizer
* #50619 [dist_optim] add distributed functional RMSprop optimizer
* #50623 [dist_optim] add distributed functional Adadelta optimizer
* **#50618 [dist_optim] add distributed functional sgd optimizer**
* #50624 [dist_optim] add distributed functional Adam optimizer

Differential Revision: [D25932778](https://our.internmc.facebook.com/intern/diff/D25932778)

wanchaol committed:
[dist_optim] add distributed functional sgd optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional sgd optimizer"




[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional sgd optimizer"




[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:
```diff
@@ -207,3 +207,5 @@ def test_dist_optim(self):
     def test_dist_optim_functional(self):
         self._test_dist_optim_base(optim.Adagrad, lr=0.05)
         self._test_dist_optim_base(optim.Adam, lr=1e-2, amsgrad=True)
+        self._test_dist_optim_base(optim.SGD, lr=0.05)
```
Is this test now duplicated with the one on line 203 of this file?

wanchaol replied:
Yes you are right, right now we are defaulting to use the torchscripted version of optimizer, let me merge these two.

rohan-varma reviewed (approved):
Looks good as it matches the implementation in https://github.com/pytorch/pytorch/pull/45597/

wanchaol committed:
Update on "[dist_optim] add distributed functional sgd optimizer"


Differential Revision: [D25932778](https://our.internmc.facebook.com/intern/diff/D25932778)

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional sgd optimizer"


Differential Revision: [D25932778](https://our.internmc.facebook.com/intern/diff/D25932778)

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional sgd optimizer"


Differential Revision: [D25932778](https://our.internmc.facebook.com/intern/diff/D25932778)

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional sgd optimizer"


Differential Revision: [D25932778](https://our.internmc.facebook.com/intern/diff/D25932778)

[ghstack-poisoned]
